samples/99_zil_interpreter/app/builtins.rb

```diff
@@ -0,0 +1,288 @@
+  arguments[0] > arguments[1]
+}
+
+# <L? ...>
+ZIL_BUILTINS[:L?] = define_for_evaled_arguments { |arguments|
+  raise FunctionError, "L? only supported with 2 arguments!" if arguments.length != 2
+  arguments[0] < arguments[1]
+}
+
+# <NOT ...>
+ZIL_BUILTINS[:NOT] = define_for_evaled_arguments { |arguments|
+  raise FunctionError, "NOT only supported with 1 argument!" if arguments.length != 1
+  arguments[0] == false
+}
+
+# <AND ...> (FSUBR)
+ZIL_BUILTINS[:AND] = lambda { |arguments, context|
+  result = false
+  arguments.each { |a|
+    result = eval_zil(a, context)
+    return false if result == false
+  }
+  result
+}
+
+#! shouldn't those just be doable with `Enumerable#all?`
+# <AND? ...> (SUBR)
+ZIL_BUILTINS[:AND?] = define_for_evaled_arguments { |arguments|
+  result = false
+  arguments.each { |a|
+    result = a
+    return false if a == false
+  }
+  result
+}
+
+# or with `Enumerable#any?` in this case
+# <OR ...>  (FSUBR)
+ZIL_BUILTINS[:OR] = lambda { |arguments, context|
+  arguments.each { |a|
+    result = eval_zil(a, context)
+    return result unless result == false
+  }
+  false
+}
+
+# <OR? ...>  (SUBR)
+ZIL_BUILTINS[:OR?] = define_for_evaled_arguments { |arguments|
+  arguments.each { |a|
+    return a unless a == false
+  }
+  false
+}
+
+# <COND () ...> (FSUBR)
+ZIL_BUILTINS[:COND] = lambda { |arguments, context|
+  # COND goes walking down its clauses, EVALing the first element of each clause,
+  # looking for a non-FALSE result. As soon as it finds a non-FALSE, it forgets
+  # about all the other clauses and evaluates, in order, the other elements of
+  # the current clause and returns the last thing it evaluates.
+  #
+  raise FunctionError, "COND requires at least one parameter!" unless arguments.length > 0
+
+  i = 0
+  while i < arguments.length
+    clause = arguments[i]
+
+    raise FunctionError, "Arguments to COND must be of List type!" unless clause.class == Syntax::List
+    raise FunctionError, "COND clauses require at least 1 item!" unless clause.elements.length > 0
+
+    clause_eval = eval_zil(clause.elements[0], context)
+    if clause_eval != false # eval 0th element for non-false
+      clause.elements.drop(1).each { |clause_element|
+        clause_eval = eval_zil(clause_element, context)
+      }
+      return clause_eval
+    end
+    i += 1
+  end
+
+  false
+}
+
+ZIL_BUILTINS[:OBJECT] = lambda { |arguments, context|
+  # Objects have a name, and a list of properties and values
+
+  raise FunctionError, "OBJECT requires at least a name!" unless arguments.length > 0
+
+  # not sure if this is needed
+  raise FunctionError, "OBJECT requires properties" unless arguments.length > 1
+
+  object_name, *object_properties = arguments
+
+  object = { name: object_name, properties: {} }
+
+  object_properties.each do |property|
+    raise FunctionError, "Parameters to object must be in list form" unless property.class == Syntax::List
```

Comment:
You'll want to look at examples in the ZIL source that contain comments and make sure your code can handle those.

```
<OBJECT TROPHY-CASE	;"first obj so L.R. desc looks right."
	(IN LIVING-ROOM)
	(SYNONYM CASE)
	(ADJECTIVE TROPHY)
	(DESC "trophy case")
	(FLAGS TRANSBIT CONTBIT NDESCBIT TRYTAKEBIT SEARCHBIT)
	(ACTION TROPHY-CASE-FCN)
	(CAPACITY 10000)>
```

Reply:
If you haven't, it's a good idea to look over the 1dungeon.zil code from the historical source zork1 code.  You may have already - it's hard to tell in this PR because of all of the extra commits.

Reply:
this will not handle comments I believe (I may be wrong though)